packages/amplify_flutter/test/config/auth_config_test.dart

```diff
@@ -39,12 +39,6 @@ void main() {
             final expectedConfig = expected[name]!;
             final cognitoConfig = config.auth!.awsPlugin!.auth!.default$!;
             expect(cognitoConfig, equals(expectedConfig));
-            expect(
```

Comment:
This was essentially testing that deserializing and then serializing the config results in the original config. That is no longer the case for older versions of the config with `loginMechanism` and `loginMechanisms` essentially being ignored.

We could have a modified version of this test where we only check a subset of the values, but I am not sure how much value that adds. I think the expectation on line 41 is sufficient.

Reply:
This may not be true for old versions but **must** be true for all new ones if we drop support for old versions. We need to just the files not passing and leave this expectation.

Reply:
> must be true for all new ones
Do you mind elaborating on that a little? I might be missing some context on this. I don't see `AmplifyConfig.toJson()` used outside of tests. I am fine with removing the tests under "old" and adding this expectation back, but would like to understand the value of this expectation and how you envision these tests working moving forward.

The latest CLI version is 7.6.3. 7.6.3 might not produce the same auth config as 6.5.0, but let's just say hypothetically that it does and these tests are effectively covering all versions of the CLI we support for the Authenticator. What happens when 8.0.0 introduces a new property. Let's say this new property is used for something outside the authenticator. We don't necessarily care about this property, but we should update `CognitoAuthConfig` to add the new property. That will make the tests for 6.5.0 fail. Won't we still want tests for 6.5.0-7.X.X in addition to tests for 8.0.0? I am envisioning that we maintain tests for every major version that we support (or every major version that produces a unique config at least).

FWIW I left the tests under "old" because I thought it would be good to have tests that validate that old configs still produce a valid `CognitoAuthConfig`. Right now we only use `CognitoAuthConfig`/`AmplifyConfig` for the authenticator, which will support 6.5.0 and up. However, if we were to start using the config for anything else, we may want to support older versions, so knowing that they produce an expected result would be valuable. With this PR, we are dropping support for anything older than 6.5.0 for the Authenticator, but I am not necessarily saying that across the board amplify_flutter does not support older versions on the config... We can always add tests for older configs if/when we use the config for other needs though so I am fine to remove the pre 6.5.0 tests.

Reply:
The utility of the `AmplifyConfig` class is two-fold for the end-user: to be able to create Amplify configurations at runtime or to be able to update Amplify configurations at runtime. In both cases, they will call `toJson` and pass the results to the `Amplify.configure`. This JSON **must** exactly match the JSON that would have been created by the CLI for the same configuration - no information can be lost or incorrectly manipulated by our decode/encode process.

That's also why support for CLI versions is all or nothing - we cannot decode an old config, then re-encode it with missing values (like loginMechanisms) - that completely defeats the purpose. Maybe not for the Authenticator, but for the utility of this class, which is public.

Reply:
If you want to keep support for decoding old configs for the purpose of the Authenticator only, then those tests should be moved to the Authenticator package, as they are no longer testing the config class.

Reply:
But I also believe we should keep login mechanisms for these reasons also - it's a very lightweight addition to the class which allows for complete backwards compatibility and a fully correct `AmplifyConfig` implementation.

Reply:
> The utility of the AmplifyConfig class is two-fold for the end-user: to be able to create Amplify configurations at runtime or to be able to update Amplify configurations at runtime. 

Why would a customer that is using `AmplifyConfig` to create or update Amplify configurations at runtime want to add a property that the amplify libraries do not use?

> In both cases, they will call toJson and pass the results to the Amplify.configure. This JSON must exactly match the JSON that would have been created by the CLI for the same configuration

For what versions of the CLI must this match? 

What about versions of the CLI between 6.5.0 and 7.6.3? Do we know that none of those versions temporarily added an attribute only for it to be removed later? Would we need to retain those? 

What about hand edited configs? If a developer adds `"foo": "bar"` to their config, does that need to be retained? 

Why would either of these hypothetical situations be any different that `loginMechanisms`? `loginMechanisms` is unused. It was never used by a stable version of any of the amplify flutter libraries (correct me if that is wrong). Why is it different than `foo`?

I know these are somewhat ridiculous hypotheticals, but I don't understand why `AmplifyConfig.fromJson(config).toJson()` has to equal the original config, and I am also not sure it is always is true right now even with `loginMechanisms`. The only way it would always equal the original config is if we maintained a copy of the original config that was passed in.

Reply:
You can make no assumptions about how the iOS or Android libraries use those config files. Can you say with 100% certainty that `{"loginMechanisms": null}`, `{"loginMechanisms": []}`,  and `{}` will all be interpreted the same way like they are in Authenticator? Now multiply that by all the values, and appreciate the fact that null de-references can and do crash programs. If there's an empty array in the JSON we parse, it should be there when it leaves - it's that simple.

I don't think customers' extraneous K-V pairs is the same thing as what I'm describing.

Reply:
By the way - this is *not* straightforward to implement and required fine-tuning the json_serializable settings. Hence the need for the unit test - I was getting null dereferences!!

Reply:
> You can make no assumptions about how the iOS or Android libraries use those config files

100% agreed. Which is why we shouldn't pass the value of AmplifyConfig.toJson() to iOS and Android. Are we doing that now? If so, I think that is an issue. We should always pass the unaltered version. If we pass the value of AmplifyConfig.toJson() to iOS and Android, we will have to really closely monitor changes in the CLI library and Amplify-iOS and Amplify-Android, regardless of wether or not we make the changes in this PR. I don't feel comfortable  passing that to iOS and Android unless we have integration tests that call configure (through to iOS and Android) with every major version of the CLI we intend to support.

Reply:
This class is not for Authenticator. It's a commitment to providing a fully-typed CLI config to customers which they can create from scratch, modify, and send to iOS/Android. This is a feature request we continuously get! The fact that Authenticator uses this is irrelevant.

Explain the difference between my unit test verifying decode/encode and an integration test calling configure and I'll be happy to implement it.

Reply:
I've already scaffolded the functionality to stay up to date with the CLI - the Node script uses the same package used internally to generate the configuration.

Reply:
Can you let me know if the value from AmplifyConfig.toJson() is currently being passed to Amplify-iOS or Amplify-Android?

Reply:
No - but again, you've missed the point. I wrote this class specifically so that you can call `.toJson` on it and pass it to iOS/Android, because that is what customers want. They don't want a half-baked AmplifyConfig object with a bunch of caveats like you can only read values from it.

Reply:
I very clearly am missing the point still 🙂 

Is there any use case (current or future) for calling `AmplifyConfig.fromJson(myJsonConfig).toJson()`?

Reply:
If we want to allow customers to call `AmplifyConfig({prop1: value1, ... })` and then call `.toJson()` and pass it to iOS and Android, that is one thing. Taking a JSON config, converting it to an AmplifyConfig object, and then converting it back to JSON to pass to iOS and Android is something different. If we take a JSON object, convert it to an AmplifyConfig object, and then back to JSON, there is no way to know that we aren't dropping values unless we test every possible combination (every possible output from every possible version of the CLI we want to support). If a customer provides a JSON config, we should just pass that straight to native. No need to serialize/deserialize.

Reply:
> Is there any use case (current or future) for calling AmplifyConfig.fromJson(myJsonConfig).toJson()?

Yes.

- https://github.com/aws-amplify/amplify-flutter/issues/1115
- https://github.com/aws-amplify/amplify-flutter/issues/132
- https://github.com/aws-amplify/amplify-flutter/issues/1090
- https://github.com/aws-amplify/amplify-flutter/issues/1208

Customers should be able to do the following:

```dart
final json = jsonDecode(amplifyconfig) as Map<String, Object?>;
final config = AmplifyConfig.fromJson(json).copyWith(...);
await Amplify.configure(jsonEncode(config));
```

> there is no way to know that we aren't dropping values unless we test every possible combination

That's not correct - you only need to check the different types (string, num, list, etc) and null/non-null/empty - which is like 15 combos - plus the AWS-specific stuff, which follows patterns is already covered at 100%. I would argue that we have close to 100% confidence with the current version range tested (CLI v5-7) that any given config will serialize correctly. 

This is a finite space we're dealing with here. I've read the iOS mobile client's auth config parsing code and looked at all the code paths to determine this info - I'm not just pulling it out of my rear.

Reply:
> I've read the iOS mobile client's auth config parsing code
Are you going to read through it every time we bump versions of amplify-ios? That isn't really a testing strategy. I'll talk to you about this one offline.